### ISSUE TITLE
Keras 2 bug in `concatenate()` in a loop appending to a tensor list (ex: densenet)

### ISSUE THREAD

**Issue by ahundt** [fchollet/keras#5972]
I've got code that works in the keras-1 compatibility layer of keras-2, but fails when I use the new `concatenate()` API.

Key lines:
```python
    concat_axis = 1 if K.image_data_format() == 'channels_first' else -1

    x_list = [x]

    for i in range(nb_layers):
        x = __conv_block(x, growth_rate, bottleneck,
                         dropout_rate, weight_decay)
        x_list.append(x)

        # commenting merge() and uncommenting concatenate() fails
        x = merge(x_list, mode='concat', concat_axis=concat_axis)
        # x = concatenate(x_list, concat_axis)

        if grow_nb_filters:
            nb_filter += growth_rate

    if return_concat_list:
        return x, nb_filter, x_list
    else:
        return x, nb_filter
```

Error:
```
Traceback (most recent call last):
  File "/Users/athundt/source/keras-contrib/keras_contrib/applications/densenet.py", line 712, in <module>
    dn = DenseNet()
  File "/Users/athundt/source/keras-contrib/keras_contrib/applications/densenet.py", line 136, in DenseNet
    model = Model(inputs, x, name='densenet')
  File "/usr/local/lib/python2.7/site-packages/Keras-2.0.1-py2.7.egg/keras/legacy/interfaces.py", line 87, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python2.7/site-packages/Keras-2.0.1-py2.7.egg/keras/engine/topology.py", line 1704, in __init__
    str(layers_with_complete_input))
RuntimeError: Graph disconnected: cannot obtain value for tensor Tensor("conv2d_2/convolution:0", shape=(?, 32, 32, 12), dtype=float32) at layer "concatenate_1". The following previous layers were accessed without issue: ['input_1', 'initial_conv2D', 'batch_normalization_1', 'activation_1', 'conv2d_1']
```

The full code is in https://github.com/farizrahman4u/keras-contrib/pull/46/commits/bf95474bdd48a85df1c443db0fa108bf28dcf86a

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by dbysll** [comment]
Similar thing happen to me.... Well, it used to work for me using 
merge([up6, conv4], mode='concat', concat_axis=1)

but after updating keras, it fails using 
concatenate([up6, conv4], axis=1)

The error message is:
  File "unet_2d.py", line 119, in get_2D_unet
    up6 = concatenate([up6, conv4], axis=1)
  File "/data/deep/anaconda/anaconda2/envs/tensorflow/lib/python2.7/site-packages/keras/layers/merge.py", line 391, in concatenate
    return Concatenate(axis=axis, **kwargs)(inputs)
  File "/data/deep/anaconda/anaconda2/envs/tensorflow/lib/python2.7/site-packages/keras/engine/topology.py", line 554, in __call__
    output = self.call(inputs, **kwargs)
  File "/data/deep/anaconda/anaconda2/envs/tensorflow/lib/python2.7/site-packages/keras/layers/merge.py", line 166, in call
    return K.concatenate(inputs, axis=self.axis)
  File "/data/deep/anaconda/anaconda2/envs/tensorflow/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 1525, in concatenate
    return tf.concat([to_dense(x) for x in tensors], axis)
  File "/data/deep/anaconda/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/ops/array_ops.py", line 1075, in concat
    dtype=dtypes.int32).get_shape(
  File "/data/deep/anaconda/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 669, in convert_to_tensor
    ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
  File "/data/deep/anaconda/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/framework/constant_op.py", line 176, in _constant_tensor_conversion_function
    return constant(v, dtype=dtype, name=name)
  File "/data/deep/anaconda/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/framework/constant_op.py", line 165, in constant
    tensor_util.make_tensor_proto(value, dtype=dtype, shape=shape, verify_shape=verify_shape))
  File "/data/deep/anaconda/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/framework/tensor_util.py", line 367, in make_tensor_proto
    _AssertCompatible(values, dtype)
  File "/data/deep/anaconda/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/framework/tensor_util.py", line 302, in _AssertCompatible
    (dtype.name, repr(mismatch), type(mismatch).__name__))
TypeError: Expected int32, got list containing Tensors of type '_Message' instead.

Please help me!!!!

**Comment by kemaswill** [comment]
@dbysll Please update your pip, tensorflow and keras respectively:
`pip install pip --upgrade`
`pip install tensorflow --upgrade`
`pip install keras --upgrade`

**Comment by dbysll** [comment]
@kemaswill  Thanks for your comments. But I believe I am already using the latest version of Keras and tensorflow.

**Comment by ahundt** [comment]
@farizrahman4u any thoughts on this? I know you were recently working on merge.py in #5812 

I can also duplicate this problem in [keras-contrib cifar10_densenet.py](https://github.com/farizrahman4u/keras-contrib/blob/master/examples/cifar10_densenet.py ), if any the `merge()` calls are appropriately changed to `concatenate()`.

**Comment by farizrahman4u** [comment]
Please provide a minimal standalone script which I can run locally.

**Comment by ahundt** [comment]
@farizrahman4u I added a pull request that reproduces the bug, see https://github.com/fchollet/keras/pull/6034

**Comment by farizrahman4u** [comment]
@ahundt Again, I don't understand how it is reproducible.. (what is __conv_block ?). Can you provide a minimal standalone script that I can quickly copy-paste to my terminal? Thanks!

**Comment by ahundt** [comment]
@farizrahman4u sorry there was a typo which is now fixed.

**Comment by ahundt** [comment]
Here is the code:

```python

    x3 = np.random.random((1, 1, 1))
    nb_layers = 4
    x_i = layers.Input(shape=(None, None))
    x_list = [x_i]
    x = x_i
    for i in range(nb_layers):
        x_list.append(x)
        # The commented line fails if uncommented
        # x = layers.concatenate(x_list)
        x = layers.merge(x_list, mode='concat', concat_axis=1)
    concat_model = models.Model(x_i, x)
    concat_out = concat_model.predict([x3])
    x3 = np.repeat(x3, 16, axis=1)
    assert concat_out.shape == (1, 16, 1)
    assert_allclose(concat_out, x3)
```

**Comment by farizrahman4u** [comment]
@ahundt Bug located.. #5972

In the mean time, You can change  `x = concatenate(x_list, axis=1)` to  `x = concatenate(x_list[:], axis=1)`.

**Comment by ahundt** [comment]
Cool, thanks! In my real use case I also need `concat_list[1:]`, so would I then do `(concat_list[1:])[:]` to get around it?

**Comment by ahundt** [comment]
This isn't for my use case, but if I'm reading the fix correctly for cases other than `axis=1` and `axis=0`, will it work for an arbitrary `axis=n` and it is a list with `m` dimensions s.t. `n<m`?

**Comment by ahundt** [comment]
Bugs are fixed & merged via #6034 and #6035
